### DISCUSSION TITLE
Custom profiling of Cull/Draw operations

### DISCUSSION THREAD

**Discussion by fteppe** [openscenegraph/OpenSceneGraph#1233] [Q&A] [answered]
Hello, Thank you for the fantastic framework.

I am in a situation where I want to use [Tracy](https://github.com/wolfpld/tracy) in an OSG app. And I would like to add profile marks on cull and render operations. But those are started and ended in OSG code. So there is no scope where I can put my profile mark. I can profile individual cull operations as long as they are overriden by user code. But I can't profile the cull operation of the entire viewer.

Is there a way I can have user code that encloses the start and end of Cull/Draw ? This is a context where they are done in a thread different from the main thread. **Basically I would like to be able to make the same measures as the ones done by the OSG profiler**.

these ones in particular:

![image](https://github.com/openscenegraph/OpenSceneGraph/assets/16321628/e7ad79ab-8151-45d6-a729-f4d53f53b399)

So either have a function that calls the actual Cull/Draw functions from the viewer where I can put my scope. Or a callback that is called when a frame cull is started, and another one that is called when it is over, so I can trigger the start and end of a scope.

The reason I want to be able to do this is because tracy offers a better visualization of nested scopes, that way it would be more easy for me to see which parts of the draw/cull operations are slow by profiling individual functions. But I also want to be able to get the more global measure to have it all in one place, the more detailed function profiling and the more high level "Cull" and "Draw" times.

**Comment by robertosfield** [comment]
The osg::Stats class is what collects the frame/update/cull/draw/draw GPU times.  I don't know if you could add anything around vsg::Stats.

The other approach would be to use a update callback, cull and draw callbacks to add in the Tracy calls.

**Comment by fteppe** [comment]
Is there any way I can make sure an update/cull/draw callback is either called first or last ?

**Comment by robertosfield** [comment, marked as answer]
node callbacks in the OSG have responsibility for traversal of the node's children so you can do timing on before traversing the subgraph and after,  Look for examples that use callbacks and discussions about it to explain the approach.

**Comment by fteppe** [comment]
Thank you!

**Comment by robertosfield** [comment]
If you care enough about performance optimization that you are thinking there is value in using Tracy then it might be that you should just move over to using the Vulkan/VulkanSceneGraph rather than OpenGL/OpenSceneGraph, this will likely give you the biggest improvement in performance you could get.

**Comment by fteppe** [comment]
It's for openMW. So VSG is a long term goal, but something that will probably not happen in the near future. And because we are adding a lot of stuff from oblivion skyrim etc... I would like to have better profiling tools on a shorter timeframe.

**Comment by Modenbach** [comment]
Hi all,

I also try getting timing information from various cameras located inside the scene.

What I do is:
         _camera->setStats(myOsgStats);
         _camera->setNodeMask(0xffffffff);
         _camera->setFinalDrawCallback(new Cl_3D_FinalDrawStatsCallback(_camera));

The callback collects the attributes times and prints them out.
Unfortunately they are all 0.0
Obviously the camera doesn't recognize the presence of the Stats instance and thus doesn't collect anything.

What am I missing to get it work?

Thanks

Werner

Am 31.05.2023 um 09:53 schrieb Robert Osfield:

**Reply by fteppe** [reply]
So my solution was much different to that, I created two cameras. 
One with pre pass, and the other with a post pass:

```cplusplus
        {
            osg::ref_ptr<osg::Camera> camera = new osg::Camera;
            camera->setRenderOrder(osg::Camera::PRE_RENDER, std::numeric_limits<int>::lowest());
            root->addChild(camera);
            camera->setClearMask(0);
            camera->setInitialDrawCallback(new CameraCallbackStart(tracyCtx, cpuScopeName, sourceLocation));
        }
        {
            osg::ref_ptr<osg::Camera> camera = new osg::Camera;
            camera->setRenderOrder(osg::Camera::POST_RENDER, std::numeric_limits<int>::max());
            root->addChild(camera);
            camera->setClearMask(0);
            camera->setFinalDrawCallback(new CameraCallBackEnd(tracyCtx));
        }
```

And in those callbacks I either started the measure or ended it. That's to measure draw and GPU.

For cull and update, I just did my measures in cull and update callbacks.
